WIP: BXMSDOC-2232-master: Update DRL rule language reference and single-source in enterprise.

sterobin opened the pull request:
See [JIRA](https://issues.jboss.org/browse/BXMSDOC-2232) for details.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-THEN-proc.adoc:
```diff
@@ -14,78 +14,56 @@ The `then` part of the rule contains the actions to be performed when the condit
 ----
 update( LoanApplication ) {
         setAmount( 100 )
 }
```
The syntax of the 'update' looks identical to the one of the 'modify' here and indeed it is wrong. Please change this code snippet with:

> loanApplication.setAmount( 100 );
> update( loanApplication );

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-THEN-proc.adoc:
```diff
@@ -14,78 +14,56 @@ The `then` part of the rule contains the actions to be performed when the condit
+[source]
 ----
-insertLogical( new Applicant() );
+delete( Applicant() );
```
This should be without parenthesis:

`delete( applicant );`

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-WHEN-proc.adoc:
```diff
@@ -11,154 +11,149 @@ The `when` part of the rule contains the conditions that must be met to execute
 . In the DRL designer, enter `when` within the rule to begin adding condition statements. The `when` section consists of zero or more fact patterns that define conditions for the rule.
 +
-If the `when` section is empty, then actions in the `then` section are executed every time a `fireAllRules()` call is made in the {DECISION_ENGINE}. This is useful if you want to use rules to set up the {DECISION_ENGINE} state.
+If the `when` section is empty, then the conditions are considered to be always true and the actions in the `then` section are executed every time a `fireAllRules()` call is made in the {DECISION_ENGINE}. This is useful if you want to use rules to set up the {DECISION_ENGINE} state.
```
This is wrong. If the when is empty the consequence is executed only the first time when a `fireAllRules()` is called on a given session, not every time.

sterobin replied:
Done. Changed to "...are considered to be true and the actions in the `then` section are executed the first time a `fireAllRules()` call ..."

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-WHEN-proc.adoc:
```diff
@@ -11,154 +11,149 @@ The `when` part of the rule contains the conditions that must be met to execute
 ----
 +
-* `collect`: Use this to define a collection of objects that the construct can use as part of the condition. In the example, all pending applications in the {DECISION_ENGINE} for each given mortgage are grouped in `ArrayLists`. If three or more pending applications are found, the rule is executed.
+* `collect`: Use this to define a collection of objects that the rule can use as part of the condition. In the example, all pending applications in the {DECISION_ENGINE} for each given mortgage are grouped in an `ArrayList`. If three or more pending applications are found, the rule is executed.
```
Please replace `ArrayList` with `List`. We want users to pattern match against the List interface instead of the concrete class that at the moment is an ArrayList but may change in future.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-WHEN-proc.adoc:
```diff
@@ -11,154 +11,149 @@ The `when` part of the rule contains the conditions that must be met to execute
-    from collect( LoanApplication( Mortgage == m, status == 'pending' ) )
+$m : Mortgage()
+$a : ArrayList( size >= 3 )
+    from collect( LoanApplication( Mortgage == $m, status == 'pending' ) )
```
Use `List`.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-central-create-proc.adoc:
```diff
@@ -127,28 +135,96 @@ rule "Using a declared type"
 ----
 +
-* `rule`: Use this to define each rule in the DRL file. Rules consist of a rule name in the format `rule "name"`, followed by optional attributes that define rule behavior (such as `salience` or `no-loop`), followed by `when` and `then` definitions. The same rule name cannot be used more than once in the same package. The `when` part of the rule contains the conditions that must be met to execute ...
+* `global`: (optional) Use this to include a global variable to be used by rules in the DRL file. Global variables typically provide data or services for the rules, such as application services used in rule consequences, and return data from rules, such as logs or values added in rule consequences. Set the global value in the working memory of the {DECISION_ENGINE} through a REST or `kmodule.xml` ...
```
It is not possible to set the value of a global in the `kmodule.xml` file.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-con.adoc:
```diff
@@ -3,4 +3,56 @@
+----
+rule "Underage"
+  salience 15
+  dialect "mvel"
```
Please let remove usage of `dialect "mvel"` wherever is possible like here. We want to deprecate and discourage it.

sterobin replied:
Sure, done. Throughout. Just FYI though, that today it still says it by default (it seems) in our BC example projects.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-con.adoc:
```diff
@@ -3,4 +3,56 @@
+end
+----
+
+A DRL file can contain single or multiple rules, queries, and functions, and can define resource declarations such as imports, globals, and attributes that are assigned and used by your rules and queries. If a DRL file contains a single rule, you can optionally use the `.rule` file extension. In large and complex rule systems, single-rule files may be processed more efficiently than files with ma ...
```
I'm a bit surprised, from where did you take this? I don't think we ever supported a `.rule` file extension. All the extensions we support are listed here https://github.com/kiegroup/droolsjbpm-knowledge/blob/master/kie-api/src/main/java/org/kie/api/io/ResourceType.java#L73

sterobin replied:
Ah, that's funny. That is mentioned in the Drools doc today, in the "A rule file" paragraph: https://docs.jboss.org/drools/release/latestFinal/drools-docs/html_single/#_a_rule_file  I'll fix as part of all of your feedback.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/property-change-listeners-con.adoc:
```diff
@@ -0,0 +1,133 @@
+
+By default, the {DECISION_ENGINE} does not re-evaluate all fact patterns for fact types each time a rule is triggered, but instead reacts only to modified properties that are constrained or bound inside a given pattern. For example, if a rule calls `modify()` as part of the rule actions but the action does not generate new data in the KIE base, the {DECISION_ENGINE} does not automatically re-eval ...
+
+NOTE: This property reactivity behavior does not apply to `update()` calls in rules, only to `modify()` calls.
```
You can remove this note. `update` and `modify` behave exactly in the same way in this regards in latest drools versions.

sterobin replied:
When you get to the "Supported rule action methods in DRL" section, you'll see modify and update there listed and explained. If they behave in exactly the same way, can you comment at that point what might need to change in the explanations, and confirm if modify is still preferable over update, and why? Sounds like that might be different now. Thanks.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-declarations-without-metadata-con.adoc:
```diff
@@ -0,0 +1,83 @@
+----
+
+When you declare a new fact type, the {DECISION_ENGINE} generates bytecode at compile time that implements a Java class representing the fact type. The generated Java class is a one-to-one JavaBeans mapping of the type definition.
+
```
I'd remove any reference to internal implementation details like bytecode generation. I don't think this is interesting for users and moreover this is no longer true when using the executable model.

sterobin replied:
A lot of the content in this DRL doc, and the majority in the decision engine doc, are internal implementation details for the users info. It would be difficult for me to be the one to say which details should stay or get cut from docs, so I rely on SMEs or QEs to tell me. In this specific case, if I remove this paragraph, then everything after it doesn't make sense and should therefore get cut out as well, I think.

Let me know what you want me to do here specifically.

mariofusco replied:
Please just reword "generates bytecode at compile time that implements a Java class ..." into "generates at compile time a Java class ..."

sterobin replied:
Done.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-globals-con.adoc:
```diff
@@ -0,0 +1,45 @@
+endif::[]
+
+Global variables in DRL files typically provide data or services for the rules, such as application services used in rule consequences, and return data from rules, such as logs or values added in rule consequences. You set the global value in the working memory of the {DECISION_ENGINE} through a REST or `kmodule.xml` configuration, declare the global variable above the rules in the DRL file, and  ...
+
```
I already added another comment about this: it is not possible to set a value for a global through the `kmodule.xml` configuration file.

sterobin replied:
@mariofusco @baldimir, agreed. I changed to "...through a KIE session configuration or REST operation..."  Usually "REST operation" is the generic term, when we don't want to say "call" and get into the API aspect. This okay with you now?

mariofusco replied:
ok for me

baldimir replied:
ok for me

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-operators-ref.adoc:
```diff
@@ -0,0 +1,244 @@
+
+`contains`, `not contains`::
+Use these operators to verify whether a field that is a `Collection` contains or does not contain a specified value. These operators apply to `Collection` properties, but you can also use these operators in place of `String.contains()` and `!String.contains()` constraints checks.
++
```
These operators work with both java `Collection`s and arrays.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-operators-ref.adoc:
```diff
@@ -0,0 +1,244 @@
+
+`memberOf`, `not memberOf`::
+Use these operators to verify whether a field is a member of or is not a member of a `Collection` that is defined as a variable. The `Collection` must be a variable.
++
```
Same as before, this also works for arrays.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-queries-con.adoc:
```diff
@@ -1,64 +1,65 @@
+QueryResults results = ksession.getQueryResults( "people under the age of 21" );
+System.out.println( "we have " + results.size() + " people under the age  of 21" );
+
+rule "Underage"
```
This rule seems to be a leftover. I don't understand why it is here.

sterobin replied:
@mariofusco @baldimir , From my understanding, this example (and a couple others similar to it) is showing the whole rule file, starting at the `query` component, and the example is headed "Example query and query results for people under the age of 21, with a rule" for this reason. The rule that follows is part of the rule file and is related to the query, which is why it's there. I suppose I can remove it, but I see value in keeping it there to show that a rule file can contain both queries and rules, just like it can contain globals, attributes, etc., all in the same DRL file as needed. 

Let me know what you think. If we change it here, I'll need to change it a few other places as well.

mariofusco replied:
I still believe that the rule can be removed.

sterobin replied:
I still don't follow, since a `query` is just another DRL component that precedes whatever rules are listed in the DRL file. But I'll take your word for it and I've removed the rule in all 3 instances where this appears.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-queries-con.adoc:
```diff
@@ -1,64 +1,65 @@
-----
-====
 
+rule "Underage"
```
Same as above: why is this rule here?

sterobin replied:
Will change according to the above decision ^^.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-THEN-methods-ref.adoc:
```diff
@@ -0,0 +1,129 @@
+----
+update( LoanApplication ) {
+        setAmount( 100 )
+}
```
This update syntax is wrong, see one of my former comment.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-THEN-methods-ref.adoc:
```diff
@@ -0,0 +1,129 @@
+.Example rule action to delete a loan applicant object
+[source]
+----
+delete( Applicant() );
```
Must  be `delete( applicant );`

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-con.adoc:
```diff
@@ -0,0 +1,362 @@
+NOTE: DRL uses `when` instead of `if` because `if` is typically part of a procedural execution flow during which a condition is checked at a specific point in time. In contrast, `when` indicates that the condition evaluation is not limited to a specific evaluation sequence or point in time, but instead occurs continually at any time. Whenever the condition is met, the actions are executed.
+
+If the `when` section is empty, then the conditions are considered to be always true and the actions in the `then` section are executed every time a `fireAllRules()` call is made in the {DECISION_ENGINE}. This is useful if you want to use rules to set up the {DECISION_ENGINE} state.
+
```
Not every time, only the first one, see above.

sterobin replied:
Done

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-con.adoc:
```diff
@@ -0,0 +1,362 @@
+ifdef::DROOLS,JBPM,OP[]
+== Auto-boxing and primitive types
+
+Drools attempts to preserve numbers in their primitive or object wrapper form, so a variable bound to an int primitive when used in a code block or expression will no longer need manual unboxing; unlike Drools 3.0 where all primitives were autoboxed, requiring manual unboxing.
```
Does it make sense to mention something as old as Drools 3? I'd remove the last sentence.

sterobin replied:
Sure. This is a community-only bit, so left as is. Changed to "... unlike former Drools versions ..." I think the comparison is helpful to mention though, for clarity.

baldimir commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-oopath-con.adoc:
```diff
@@ -0,0 +1,158 @@
+Student( $grade: /plan?/exams[ course == "Big Data" ]?/grades )
+----
+
+Another alternative for enabling OOPath expression reactivity is to use the dedicated implementations for `List` and `Set` methods in {PRODUCT}. These implementations use specialized `ReactiveList` and `ReactiveSet` methods. A `ReactiveCollection` method is also available. The implementations also provide reactive support for performing mutable operations through the `Iterator` and `ListIterator` ...
```
The term "method" here is incorrect. It should be "Another alternative for enabling OOPath expression reactivity is to use the dedicated implementations of `List` and `Set` interfaces in {PRODUCT}. These implementations are classes `ReactiveList` and `ReactiveSet`. A `ReactiveCollection` class is also available. The implementation also provide reactive support for performing mutable operations through the `Iterator` and `ListIterator` classes."

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-oopath-con.adoc:
```diff
@@ -0,0 +1,158 @@
+
+Another alternative for enabling OOPath expression reactivity is to use the dedicated implementations for `List` and `Set` methods in {PRODUCT}. These implementations use specialized `ReactiveList` and `ReactiveSet` methods. A `ReactiveCollection` method is also available. The implementations also provide reactive support for performing mutable operations through the `Iterator` and `ListIterator` ...
+
+The following example class uses these implementations to configure OOPath expression reactivity:
```
Should be "The following example class uses these classes to configure OOPath expression reactivity"

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-oopath-con.adoc:
```diff
@@ -0,0 +1,158 @@
+    }
+  }
+----
+<1> Uses the specialized `ReactiveList` method for reactive support over the standard Java `List` method.
```
Should be: "Uses the `ReactiveList` instance for reactive support over the standard Java `List` instance."

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-oopath-con.adoc:
```diff
@@ -0,0 +1,158 @@
+----
+<1> Uses the specialized `ReactiveList` method for reactive support over the standard Java `List` method.
+<2> Uses the required `notifyModification()` method for when a field is changed in reactive support.
+<3> Due to the `ReactiveList` method, the `notifyModification()` method is not required. The notification is handled automatically, like all other mutating operations performed over the `children` field.
```
Should be: "Because the children field is a `ReactiveList` instance, the `notifyModification()` method call is not required. The notification is handled automatically, like all other mutating operations performed over the `children` field."

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-THEN-proc.adoc:
```diff
@@ -14,78 +14,56 @@ The `then` part of the rule contains the actions to be performed when the condit
 ----
 +
-* `insert`: Use this to insert a `new` fact and define resulting fields and values as needed for the fact.
+* `insertLogical`: Use this to insert a `new` fact logically into the {DECISION_ENGINE} and define resulting fields and values as needed for the fact. The {DECISION_ENGINE} is responsible for logical decisions on insertions and retractions of facts. After regular or stated insertions, facts must be retracted explicitly. After logical insertions, the facts that were inserted are automatically retr ...
```
Please remove "and define resulting fields and values as needed for the fact" from the first sentence. Also please extend "when the conditions that inserted the facts are no longer true" in the last sentence to "when the conditions (a rule) that inserted the facts are no longer true".

sterobin replied:
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-con.adoc:
```diff
@@ -3,4 +3,56 @@
 
 DRL (Drools Rule Language) rules are business rules that you define directly in `.drl` text files. These DRL files are the source in which all other rule assets in {CENTRAL} are ultimately rendered. You can create and manage DRL files within the {CENTRAL} interface, or create them externally using Red Hat Developer Studio, Java objects, or Maven archetypes. A DRL file can contain one or more rule ...
```
Agreed, done. But tweaked to "or create them externally as part of a Maven or Java project using Red Hat Developer Studio or another integrated development environment (IDE)"

sterobin commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-con.adoc:
```diff
@@ -3,4 +3,56 @@
+
+A DRL file can contain single or multiple rules, queries, and functions, and can define resource declarations such as imports, globals, and attributes that are assigned and used by your rules and queries. If a DRL file contains a single rule, you can optionally use the `.rule` file extension. In large and complex rule systems, single-rule files may be processed more efficiently than files with ma ...
+
+Each rule must have a unique name within the rule package. If you use the same rule name twice in the same DRL file, the rules fail to compile. If you add a DRL file that includes a rule name already in the package, the new version of the rule replaces the existing version of the rule. Always enclose rule names with double quotation marks (`rule "rule name"`) to  prevent possible compilation erro ...
```
@baldimir I think you're correct, actually. Even if not, it's safer to just say to users to never use the same name, either way. So I've changed to "If you use the same rule name more than once in any DRL file in the package, the rules fail to compile." Just to cover our bases.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-packages-con.adoc:
```diff
@@ -0,0 +1,26 @@
+[id='drl-packages-con_{context}']
+= Packages in DRL
+
+A package is a folder of related assets in {PRODUCT}, such as data objects, DRL files, decision tables, and other asset types. A package also serves as a unique namespace for each group of rules. You can assemble rules from multiple rule sources with the same top-level package configuration, but you cannot merge resources declared under different names into the same package. However, a single rul ...
```
@baldimir Yeah, I see what you mean about that sentence in particular. I do see value in some of the other content, though, to emphasize what the package must contain for declaring it in a DRL file.

I've revised it to this. Wdyt?

>A package is a folder of related assets in {PRODUCT}, such as data objects, DRL files, decision tables, and other asset types. A package also serves as a unique namespace for each group of rules. A single rule base can contain multiple packages. The package that you declare in a DRL file must be the package where all of the assets related to the DRL file are contained. A common approach is to store all the rules for a package in the same file as the package declaration so that the package is entirely self-contained.

baldimir replied:
"The package that you declare in a DRL file must be the package where all of the assets related to the DRL file are contained. " This is incorrect. You can import any asset you need to a DRL file with "import" keyword.

sterobin replied:
Done, reworded to "... multiple packages. You typically store all the rules for a package in the same file as the package declaration so that the package is self-contained. However, you can import objects from other packages that you want to use in the rules."

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-packages-con.adoc:
```diff
@@ -0,0 +1,26 @@
+.Example package definition in a DRL file
+[source]
+----
+package mortgages.mortgages;
```
Sure, done. I used this because that's the package path/name in BC. I updated throughout, except in other docs that refer specifically to BC.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-imports-con.adoc:
```diff
@@ -0,0 +1,17 @@
+image::LanguageReference/import.png[align="center"]
+endif::[]
+
+Similar to import statements in Java, imports in DRL files identify the fully qualified paths and type names for any objects that you want to use in the rules. You specify the package and data object in the format `packageName.objectName`, one import per line. The {DECISION_ENGINE} automatically imports classes from the Java package of the same name and from the package `java.lang`.
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-imports-con.adoc:
```diff
@@ -0,0 +1,17 @@
+.Example import statement in a DRL file
+[source]
+----
+import mortgages.mortgages.LoanApplication;
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/AuthoringAssets/drl-rules-central-create-proc.adoc:
```diff
@@ -24,61 +24,69 @@ endif::[]
 ----
 +
-* `function`: (optional) Use this to include a function to be used by rules in the DRL file. Functions put semantic code in your rule source file. Functions are especially useful if an action (`then`) part of a rule is used repeatedly and only the parameters differ for each rule. Above the rules in the DRL file, you can declare the function or import a static method as a function, and then use th ...
+* `function`: (optional) Use this to include a function to be used by rules in the DRL file. Functions in DRL files put semantic code in your rule source file instead of in Java classes. Functions are especially useful if an action (`then`) part of a rule is used repeatedly and only the parameters differ for each rule. Above the rules in the DRL file, you can declare the function or import a stat ...
```
Sure, done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-functions-con.adoc:
```diff
@@ -0,0 +1,39 @@
+image::LanguageReference/function.png[align="center"]
+endif::[]
+
+Functions in DRL files put semantic code in your rule source file instead of in Java classes. Functions are especially useful if an action (`then`) part of a rule is used repeatedly and only the parameters differ for each rule. Above the rules in the DRL file, you can declare the function or import a static method in a helper class as a function, and then use the function by name in an action (`t ...
```
Ditto, done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-queries-con.adoc:
```diff
@@ -1,64 +1,65 @@
 image::LanguageReference/query.png[align="center"]
+endif::[]
 
+Queries in DRL files search the working memory of the {DECISION_ENGINE} for facts related to the rules in the DRL file. Queries search for a set of defined conditions and do not require `when` or `then` specifications. Query names are global to the KIE base and therefore must be unique among all other rule queries in the project. To return the results of a query, you construct a traditional `Quer ...
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-queries-con.adoc:
```diff
@@ -1,64 +1,65 @@
-.Query People over the age of 30
-====
-[source,java]
+.Modified query to iterate over query results
```
Done, changed to "Example iteration over query results" for consistency.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-declarations-con.adoc:
```diff
@@ -0,0 +1,15 @@
+
+Declarations in DRL files define new fact types or metadata for fact types to be used by rules in the DRL file:
+
+* *New fact types:* The default fact type in the `java.lang` package of {PRODUCT} is `Object`, but you can declare other types in DRL files as needed. Declaring fact types in DRL files enables you to define a new fact model directly in the {DECISION_ENGINE}, without creating models in a lower-level language like Java. You can also declare a new type when a domain model is already built and you wa ...
```
@baldimir @mariofusco , I hope not, because the majority of the type delcaration content that follows is all about declaring types in DRL, and is quite detailed. I vote we keep it in, if only because of how much time it took me to work through :) Jk (kind of). If we decide to go a different route, please assess the rest of the type declarations chapter as well to determine how it would change or be reduced.

Let me know what you want to do.

baldimir replied:
OK, so just leave it as it is :)

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-declarations-extended-con.adoc:
```diff
@@ -0,0 +1,23 @@
+[id='drl-declarations-extended-con_{context}']
+= Extended type declarations in DRL
+
+DRL supports extended type declarations for inheritance in the format `declare <factType1> extends <factType2>`. To extend a type declared in Java by a subtype declared in DRL, you repeat the parent type in a declaration statement without any fields.
```
Sure, was just keeping to the key word "extend/extended", but I see your point. Done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/property-change-listeners-con.adoc:
```diff
@@ -0,0 +1,133 @@
+[source,java]
+----
+@classReactive
+    public static class Person {
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/property-change-listeners-con.adoc:
```diff
@@ -0,0 +1,133 @@
+[source,java]
+----
+@propertyReactive
+    public static class Person {
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-globals-con.adoc:
```diff
@@ -0,0 +1,45 @@
+.Example global list configuration for the {DECISION_ENGINE}
+[source]
+----
+List list = new ArrayList();
```
Done, fixed throughout.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-globals-con.adoc:
```diff
@@ -0,0 +1,45 @@
+
+A use case for a global variable might be an instance of an email service. In your integration code that is calling the {DECISION_ENGINE}, you obtain your `emailService` object and then set it in the working memory of the {DECISION_ENGINE}. In the DRL file, you declare that you have a global of type `emailService` and give it the name `"email"`, and then in your rule consequences, you can use act ...
+
+You can also use the `from` keyword to pass a Hibernate session as a global, or to pull data from a named Hibernate query.
```
@baldimir @mariofusco I removed that sentence. And I don't follow why we would put that warning in this section about globals. It should go in the `then` section(s) of the doc, but we do already explain there that "Effective rule actions are small, declarative, and readable. If you need to use imperative or conditional code in rule actions, then divide the rule into multiple smaller and more declarative rules." etc etc.

Let me know what you want to do.

baldimir replied:
Thanks for removing the sentence, it is ok for me now. With the warning I just wanted to stress out that time heavy operations should not be used in rules. But based on your comment it is not needed.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-con.adoc:
```diff
@@ -0,0 +1,362 @@
+
+A _constraint_ is an expression that returns `true` or `false`. Pattern constraints in DRL are essentially Java expressions with some enhancements, such as property access, and some differences, such as `equals()` and `!equals()` semantics for `==` and `!=` (instead of the usual `same` and `not same` semantics).
+
+Any JavaBeans property can be accessed directly from pattern constraints. A bean property is exposed using a standard JavaBeans getter, such as the method `getMyProperty()` (or `isMyProperty()` for a primitive `boolean` value), which takes no arguments and returns something. For example, the `age` property is written as `age` in DRL instead of the getter `getAge()`:
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-operator-precedence-ref.adoc:
```diff
@@ -0,0 +1,73 @@
+
+DRL supports standard Java operator precedence for applicable constraint operators, with some exceptions and with some additional operators that are unique in DRL. The following table lists DRL operator precedence, where applicable:
+
+.Operator precedence in DRL pattern constraints
```
Fair point. It's precedence in the order listed. I've changed to "... lists DRL operator precedence where applicable, from highest to lowest precedence: ..."

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+
+// Prefix `and`:
+Color( colorType : type )
+and Person( favoriteColor == colorType )
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+
+// Prefix `or`:
+Color( colorType : type )
+or Person( favoriteColor == colorType )
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+--
+
+`exists`::
+Use this to specify facts and constraints that must exist. Note that this does not mean that a fact exists, but that the fact must exist. This option is triggered on only the first match, not subsequent matches. If you use this element with multiple patterns, enclose the patterns with parentheses `()`.
```
Done. Just removed that sentence.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+--
+
+`not`::
+Use this to specify facts and constraints that must not exist. Note that this does not mean that a fact does not exist, but that the fact must not exist. If you use this element with multiple patterns, enclose the patterns with parentheses `()`.
```
Done

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+--
+
+`forall`::
+Use this to set up a construct where all facts that match the first pattern match all the remaining patterns. When a `forall` construct is satisfied, the rule evaluates to `true`. This element is a scope delimiter, so it can use any previously bound variable, but no variable bound inside of it is available for use outside of it.
```
Agreed, done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+--
+
+`from`::
+Use this to specify a source for data to be matched by the conditional pattern. This enables the {DECISION_ENGINE} to reason over data that is not in the working memory. The data source can be a sub-field on a bound variable or the results of a method call. This element also enables integration with other application components and frameworks. One common example is the integration with data retri ...
```
Done, done, and done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-WHEN-elements-ref.adoc:
```diff
@@ -0,0 +1,764 @@
+end
+----
+
+.Example rule with `from` and `lock-on-active` rule attribute
```
Oops! Fixed. Copied the wrong condition block. Done. Good catch!

sterobin commented:
@mariofusco @baldimir , I've implemented your feedback, updated the doc previews, and resolved most comments. Please see the remaining comments that I've left open and replied to for your additional input. Thanks!

baldimir commented:
@sterobin thanks for the fixes, just one comment remaining from me. Please see in the code (or search for "The package that you declare in a DRL file must be the package where all of the assets related to the DRL file are contained. " in the comments here).

emmurphy1 commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-declarations-metadata-tags-ref.adoc:
```diff
@@ -0,0 +1,297 @@
+Person( lastName == "Doe"; firstName == "John", $a : age )
+----
+
+Positional arguments that contain a previously declared binding constrain against that binding using unification. These arguments are known as _input arguments_. If the binding does not yet exist, the argument generates the declaration and binds it to the field represented by the positional argument. These arguments are known as _output arguments_.
```
Don't know if you can improve it, but I find this sentence difficult to read.

emmurphy1 commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-THEN-con.adoc:
```diff
@@ -0,0 +1,19 @@
+[id='drl-rules-THEN-con_{context}']
+= Rule actions in DRL (THEN)
+
+The `then` part of the rule (also known as the _Right Hand Side (RHS)_ of the rule) contains the actions to be performed when the conditional part of the rule has been met. Actions consist of one or more _methods_ that execute consequences based on the rule conditions and on available data objects in the package. For example, if a bank requires loan applicants to have over 21 years of age (with a ...
```
Really just a comment:
'have over 21 years of age '
This seems somewhat stilted or as if a foreign translation. For example, in French :+1: 
j'ai 21 ans (literally I have 21 years)

I assume there is a reason you don't just say 'to be at least 21 years old'. I've seen it in other related docs as well.

emmurphy1 commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rules-THEN-methods-ref.adoc:
```diff
@@ -0,0 +1,128 @@
+
+NOTE: If you provide property-change listeners, you do not need to call this method when an object changes. For more information about property-change listeners, see
+ifdef::DM,PAM[]
+{URL_DECISION_ENGINE_DOC}#property-change-listeners-con_decision-engine[_{DECISION_ENGINE_DOC}_].
```
Link broken
